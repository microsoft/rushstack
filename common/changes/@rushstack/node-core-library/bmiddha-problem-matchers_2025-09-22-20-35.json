{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `FileError.getProblemMatcher()` which returns the problem matcher compatible with `IOperationExecutionResult.problemCollector` as well as VS Code, GitHub Actions",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}
