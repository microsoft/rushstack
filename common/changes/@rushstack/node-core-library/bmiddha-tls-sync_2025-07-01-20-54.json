{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add a `Async.runWithTimeoutAsync` function that executes an async function, resolving if the specified timeout elapses first.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}