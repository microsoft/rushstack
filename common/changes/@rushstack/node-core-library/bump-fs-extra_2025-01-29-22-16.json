{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update fs-extra to 11.3.0.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}