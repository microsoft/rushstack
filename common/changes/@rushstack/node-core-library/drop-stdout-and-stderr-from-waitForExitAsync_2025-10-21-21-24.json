{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update the return type of `Executable.waitForExitAsync` to omit `stdout` and `stderr` if an `encoding` parameter isn't passed to the options object.",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}