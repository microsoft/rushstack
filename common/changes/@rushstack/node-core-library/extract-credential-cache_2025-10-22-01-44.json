{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add \"Objects.areDeepEqual\" and \"User.getHomeFolder\" APIs.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}