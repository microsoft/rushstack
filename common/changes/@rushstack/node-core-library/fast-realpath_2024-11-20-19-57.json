{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `RealNodeModulePathResolver` class to get equivalent behavior to `realpath` with fewer system calls (and therefore higher performance) in the typical scenario where the only symlinks in the repository are inside of `node_modules` folders and are links to package folders.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}