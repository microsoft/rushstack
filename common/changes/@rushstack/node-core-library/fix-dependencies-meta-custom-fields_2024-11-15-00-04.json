{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Extend IDependenciesMetaTable interface to support custom properties in \"dependenciesMeta\" entries",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}