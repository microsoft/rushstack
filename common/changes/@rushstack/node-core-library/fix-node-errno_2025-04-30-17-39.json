{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Fix a bug in `FileSystem.isErrnoException` that failed to identify errors if the underlying method was invoked using only a file descriptor, e.g. for `fs.readSync`.",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}