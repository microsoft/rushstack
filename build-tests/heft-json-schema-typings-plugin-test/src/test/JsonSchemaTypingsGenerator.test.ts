// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { FileSystem, type FolderItem, PackageJsonLookup } from '@rushstack/node-core-library';

async function getFolderItemsAsync(
  absolutePath: string,
  relativePath: string
): Promise<Record<string, string>> {
  const folderQueue: [string, string][] = [[absolutePath, relativePath]];
  const results: [string, string][] = [];
  for (const [folderAbsolutePath, folderRelativePath] of folderQueue) {
    const folderItems: FolderItem[] = await FileSystem.readFolderItemsAsync(folderAbsolutePath);
    for (const item of folderItems) {
      const itemName: string = item.name;
      const itemAbsolutePath: string = `${folderAbsolutePath}/${itemName}`;
      const itemRelativePath: string = `${folderRelativePath}/${itemName}`;
      if (item.isDirectory()) {
        folderQueue.push([itemAbsolutePath, itemRelativePath]);
      } else {
        const itemContents: string = await FileSystem.readFileAsync(itemAbsolutePath);
        results.push([itemRelativePath, itemContents]);
      }
    }
  }

  results.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  return Object.fromEntries(results);
}

describe('json-schema-typings-plugin', () => {
  it('should generate typings for JSON Schemas', async () => {
    const rootFolder: string | undefined = PackageJsonLookup.instance.tryGetPackageFolderFor(__dirname);
    if (!rootFolder) {
      throw new Error('Could not find root folder for the test');
    }

    const folderItems: Record<string, string> = await getFolderItemsAsync(
      `${rootFolder}/temp/schema-dts`,
      '.'
    );
    expect(folderItems).toMatchSnapshot();
  });
});
