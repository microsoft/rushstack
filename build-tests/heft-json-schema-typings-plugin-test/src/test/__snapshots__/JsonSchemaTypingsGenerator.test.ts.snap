// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-schema-typings-plugin should generate typings for JSON Schemas 1`] = `
Object {
  "./test-invalid-additional.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestInvalidAdditional {
  [k: string]: unknown;
}
",
  "./test-invalid-format.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestInvalidFormat {
  [k: string]: unknown;
}
",
  "./test-schema-draft-04.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestSchemaFile {
  exampleString: string;
  exampleLink?: string;
  exampleArray: string[];
  /**
   * Description for exampleOneOf - this is a very long description to show in an error message
   */
  exampleOneOf?: Type1 | Type2;
  exampleUniqueObjectArray?: {
    field2?: string;
    field3?: string;
  }[];
}
/**
 * Description for type1
 */
export interface Type1 {
  /**
   * Description for field1
   */
  field1: string;
}
/**
 * Description for type2
 */
export interface Type2 {
  /**
   * Description for field2
   */
  field2: string;
  /**
   * Description for field3
   */
  field3: string;
}
",
  "./test-schema-draft-07.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestSchemaFile {
  exampleString: string;
  exampleLink?: string;
  exampleArray: string[];
  /**
   * Description for exampleOneOf - this is a very long description to show in an error message
   */
  exampleOneOf?: Type1 | Type2;
  exampleUniqueObjectArray?: {
    field2?: string;
    field3?: string;
  }[];
}
/**
 * Description for type1
 */
export interface Type1 {
  /**
   * Description for field1
   */
  field1: string;
}
/**
 * Description for type2
 */
export interface Type2 {
  /**
   * Description for field2
   */
  field2: string;
  /**
   * Description for field3
   */
  field3: string;
}
",
  "./test-schema-invalid.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface HttpExampleComSchemasTestSchemaNestedChildSchemaJson {
  [k: string]: unknown;
}
",
  "./test-schema-nested-child.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface HttpExampleComSchemasTestSchemaNestedChildSchemaJson {
  [k: string]: unknown;
}
",
  "./test-schema-nested.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestSchemaFile {
  exampleString: string;
  exampleLink?: string;
  exampleArray: string[];
  /**
   * Description for exampleOneOf - this is a very long description to show in an error message
   */
  exampleOneOf?: Type1 | Type2;
  exampleUniqueObjectArray?: Type2[];
}
/**
 * Description for type1
 */
export interface Type1 {
  /**
   * Description for field1
   */
  field1: string;
}
/**
 * Description for type2
 */
export interface Type2 {
  /**
   * Description for field2
   */
  field2: string;
  /**
   * Description for field3
   */
  field3: string;
}
",
  "./test-schema.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestSchemaFile {
  exampleString: string;
  exampleLink?: string;
  exampleArray: string[];
  /**
   * Description for exampleOneOf - this is a very long description to show in an error message
   */
  exampleOneOf?: Type1 | Type2;
  exampleUniqueObjectArray?: {
    field2?: string;
    field3?: string;
  }[];
}
/**
 * Description for type1
 */
export interface Type1 {
  /**
   * Description for field1
   */
  field1: string;
}
/**
 * Description for type2
 */
export interface Type2 {
  /**
   * Description for field2
   */
  field2: string;
  /**
   * Description for field3
   */
  field3: string;
}
",
  "./test-valid.schema.json.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface TestValid {
  [k: string]: unknown;
}
",
}
`;
